### ISSUE TITLE
[feature/netcore] InvalidOperationException: A request container already exists on the request.

### ISSUE THREAD

**Issue by Illeris** [OData/WebApi#1217]
[feature/netcore] 
Exception in .NET CORE 2.0 (mvc & odata). 
   InvalidOperationException: A request container already exists on the request.
Every time the odata route path is accessed. The MVC routing works as it should.

Present in 7.0.0-Nightly201801191851.

` public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseBrowserLink();
            }
            else
            {
                app.UseExceptionHandler("/Home/Error");
            }
           
            app.UseStaticFiles();

            // Source: https://github.com/OData/WebApi/issues/975
            var builder = new Microsoft.AspNet.OData.Builder.ODataConventionModelBuilder(app.ApplicationServices);
            builder.EntitySet<Demo>("Demos");

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            }
            );

            app.UseMvc(routebuilder =>
            {
                
                routebuilder.MapODataServiceRoute("defaultodatatrouting", "dms", builder.GetEdmModel());
                routebuilder.EnableDependencyInjection();
            });
        }`

Every time http://localhost:<port>/dms is accessed, the exception occurs. Even using another route path does not change the exception.
Full exception dump:

An unhandled exception occurred while processing the request.
InvalidOperationException: A request container already exists on the request.
Microsoft.AspNet.OData.Extensions.HttpRequestExtensions.CreateRequestContainer(HttpRequest request, string routeName)

Stack Query Cookies Headers
InvalidOperationException: A request container already exists on the request.
Microsoft.AspNet.OData.Extensions.HttpRequestExtensions.CreateRequestContainer(HttpRequest request, string routeName)
Microsoft.AspNet.OData.Routing.ODataPathRouteConstraint.GetODataPath(string oDataPathString, string uriPathString, string queryString, Func<IServiceProvider> requestContainerFactory)
Microsoft.AspNet.OData.Routing.ODataPathRouteConstraint.Match(HttpContext httpContext, IRouter route, string routeKey, RouteValueDictionary values, RouteDirection routeDirection)
Microsoft.AspNetCore.Routing.RouteConstraintMatcher.Match(IDictionary<string, IRouteConstraint> constraints, RouteValueDictionary routeValues, HttpContext httpContext, IRouter route, RouteDirection routeDirection, ILogger logger)
Microsoft.AspNetCore.Routing.RouteBase.RouteAsync(RouteContext context)
Microsoft.AspNetCore.Routing.RouteCollection+<RouteAsync>d__9.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Builder.RouterMiddleware+<Invoke>d__4.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Builder.RouterMiddleware+<Invoke>d__4.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware+<Invoke>d__7.MoveNext()

Show raw exception details
System.InvalidOperationException: A request container already exists on the request.
   at Microsoft.AspNet.OData.Extensions.HttpRequestExtensions.CreateRequestContainer(HttpRequest request, String routeName)
   at Microsoft.AspNet.OData.Routing.ODataPathRouteConstraint.GetODataPath(String oDataPathString, String uriPathString, String queryString, Func`1 requestContainerFactory)
   at Microsoft.AspNet.OData.Routing.ODataPathRouteConstraint.Match(HttpContext httpContext, IRouter route, String routeKey, RouteValueDictionary values, RouteDirection routeDirection)
   at Microsoft.AspNetCore.Routing.RouteConstraintMatcher.Match(IDictionary`2 constraints, RouteValueDictionary routeValues, HttpContext httpContext, IRouter route, RouteDirection routeDirection, ILogger logger)
   at Microsoft.AspNetCore.Routing.RouteBase.RouteAsync(RouteContext context)
   at Microsoft.AspNetCore.Routing.RouteCollection.<RouteAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Builder.RouterMiddleware.<Invoke>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.<Invoke>d__7.MoveNext()

**Comment by Illeris** [comment]
Hmm... Strange..  This works:
`
            app.UseMvc(routes =>
            {


                routes.MapODataServiceRoute("defaultodatatrouting", "xxx", builder.GetEdmModel());
                routes.EnableDependencyInjection();

                routes.MapRoute(
                    name: "default",
                    template: "{controller=Home}/{action=Index}/{id?}");
            });
`

Or in plain English: if you first map the ODataServiceRoute, and then the MVC route, it works. Vice versa MVC works, but OData gives the exception...

**Comment by robward-ms** [comment]
Possibly related to #1175.

**Comment by robward-ms** [comment]
Indeed related to, and fixed by, #1175. In this case, the action selector was creating the route container before the route constraint, which caused the route constraint to fail. Reversing the order of the MVBC and OData route causes that relationship to flip-flop and work as expected.

**Comment by robward-ms** [comment]
Fixed in Microsoft.AspNetCore.OData 7.0.0-Nightly201802092319.

**Comment by xuzhg** [comment]
Fixed and close it. Please re-open it or file new issue if you have further concern. Thanks.

**Comment by futurify-ydang** [comment]
This issue still occur in version 7.1.0 of my project
I use VS Code to debug and the exception is
```
An unhandled exception occurred while processing the request.
InvalidOperationException: A request container already exists on the request.
DynamicEdmModelCreation.CustomODataPathRouteConstraint.Match(HttpContext httpContext, IRouter route, string routeKey, RouteValueDictionary values, RouteDirection routeDirection) in CustomODataPathRouteConstraint.cs, line 40
```

i try to resolve table's name from the url dynamically and then set to **RoutingConventionsStore**.
By this way in my odata controller i can get the name of table and return data.

Code in **CustomODataPathRouteConstraint.cs**
``````````
namespace DynamicEdmModelCreation
{
    public class CustomODataPathRouteConstraint : ODataPathRouteConstraint
    {
        public CustomODataPathRouteConstraint(string routeName)
            : base(routeName)
        {
        }

        public override bool Match(HttpContext httpContext, IRouter route, string routeKey, RouteValueDictionary values, RouteDirection routeDirection)
        {
            try
            {

                if (routeDirection == RouteDirection.IncomingRequest)
                {
                    object dataSource;
                    if (values.TryGetValue(ODataRouteConstants.ODataPath, out dataSource))
                    {
                        httpContext.Request.SetDataSource(dataSource.ToString().Split("/").First());
                    }
                }
                var isMatch = base.Match(httpContext, route, routeKey, values, routeDirection);
                return isMatch;
            }
            catch (System.Exception ex)
            {
                throw ex;
            }
            
        }
    }
}
``````````
if i use VS 2017 to debug i get another exception message `ODataUnrecognizedPathException`
I dont know how to correct the match method of **ODataPathRouteConstraint** class.
Please give me a help! I can provide more information. 
Thanks you!
